src/concurrency/vector_clock.rs

```diff
@@ -50,21 +52,39 @@ const SMALL_VECTOR: usize = 4;
 /// so that diagnostics can report what code was responsible for an operation.
 #[derive(Clone, Copy, Debug)]
 pub struct VTimestamp {
-    time: u32,
+    // The lowest bit indicates read type, the rest is the time.
+    time_and_read_type: u32,
     pub span: Span,
 }
 
 impl VTimestamp {
-    pub const ZERO: VTimestamp = VTimestamp { time: 0, span: DUMMY_SP };
+    pub const ZERO: VTimestamp = VTimestamp { time_and_read_type: 0, span: DUMMY_SP };
+
+    fn time(&self) -> u32 {
+        self.time_and_read_type.shr(1)
+    }
+
+    fn set_time(&mut self, time: u32) {
+        self.time_and_read_type =
+            (self.time_and_read_type & 1) | time.checked_shl(1).expect("Vector clock overflow")
```

Comment:
This just checks whether `1 >= 32`. What we actually want to check is whether we are losing a bit here.